test/libdnf/base/test_loggers.cpp

```diff
@@ -0,0 +1,116 @@
+// 2. Create a LogRouter instance with two StreamLogger instances attached.
+// 3. Write messages from MemoryBufferLoger instance into LogRouter instance.
+// 4. Write aditional message into LogRouter instance.
+// 5. Check content of streams of both StreamLogger instances.
+void LoggersTest::test_loggers() {
+
+    time_t msg_time = 1582604702;  // Timestamp of the first message. "2020-02-25T04:25:02Z"
+    const pid_t pid = 25;          // Process identifier.
+
+    // Text that is expected in the logs.
+    const std::string LOG_OUT =
+R"**(2020-02-25T04:25:06Z [25] INFO Info message
+2020-02-25T04:25:07Z [25] DEBUG Debug message
+2020-02-25T04:25:08Z [25] TRACE Trace message
+2020-02-25T04:25:09Z [25] CRITICAL Critical message
+2020-02-25T04:25:10Z [25] ERROR Error message
+2020-02-25T04:25:11Z [25] WARNING Warning message
+2020-02-25T04:25:12Z [25] NOTICE Notice message
+2020-02-25T04:25:13Z [25] INFO Info message
+2020-02-25T04:25:14Z [25] DEBUG Debug message
+2020-02-25T04:25:15Z [25] TRACE Trace message
+2020-02-25T04:25:16Z [25] INFO Info additional message
+)**";
+
+    // ====================
+    // 1. Write messages into a circular MemoryBufferLogger instance.
+    // ====================
+    // Create circular memory buffer logger with capacity 10 messages (4 pre-allocated from start).
+    std::unique_ptr<libdnf::MemoryBufferLogger> memory_buffer_logger =
+        std::make_unique<libdnf::MemoryBufferLogger>(10, 4);
+
+    // Write messages into memory buffer loger.
+    for (int i = 0; i < 2; ++i) {
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::CRITICAL, "Critical message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::ERROR, "Error message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::WARNING, "Warning message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::NOTICE, "Notice message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::INFO, "Info message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::DEBUG, "Debug message");
+        memory_buffer_logger->write(msg_time++, pid, libdnf::Logger::Level::TRACE, "Trace message");
+    }
+
+    // ====================
+    // 2. Create a LogRouter instance with two StreamLogger instances attached.
+    // ====================
+    // Create log router.
+    std::unique_ptr<libdnf::LogRouter> log_router = std::make_unique<libdnf::LogRouter>();
+
+    // Create two streams for logging. Save pointers to them. Pointers will be needed later in test of streams content.
+    std::unique_ptr<std::ostringstream> log1_stream = std::make_unique<std::ostringstream>();
+    std::unique_ptr<std::ostringstream> log2_stream = std::make_unique<std::ostringstream>();
+    auto log1_stream_ptr = log1_stream.get();
+    auto log2_stream_ptr = log2_stream.get();
```

Comment:
clang-tidy suggests using:
auto * log1_stream_ptr

Reply:
Hmm. My clang-tidy do not suggest this. I have older clang-tidy than you. And may be different config.
Anyway `auto` defines right type automatically. But for "better" readability I will explicitly mark the variable as pointer (adding *).